platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java

```diff
@@ -65,9 +68,45 @@ public DiagnosticToProblemListener(InternalProblemReporter problemReporter, Cont
 
     @Override
     public void report(Diagnostic<? extends JavaFileObject> diagnostic) {
+        switch (diagnostic.getKind()) {
+            case ERROR:
+                errorCount++;
+                break;
+            case WARNING:
+            case MANDATORY_WARNING:
+                warningCount++;
+                break;
+            default:
+                break;
+        }
+
         problemReporter.reporting(spec -> buildProblem(diagnostic, spec));
     }
 
+    /**
+     * This method is based on {JavaCompiler#printCount()}
+     */
+    public void reportDiagnosticCounts() {
```

Comment:
❌ I think we should emphasise the side effect of printing something to stderr in the name. 

```suggestion
    public void printDiagnosticCounts() {
```